azurelinuxagent/pa/provision/factory.py

```diff
@@ -16,19 +16,24 @@
 #
 
 import azurelinuxagent.common.conf as conf
-
+from azurelinuxagent.common import logger
 from azurelinuxagent.common.version import DISTRO_NAME, DISTRO_VERSION, \
                                      DISTRO_FULL_NAME
 
 from .default import ProvisionHandler
-from .cloudinit import CloudInitProvisionHandler
+from .cloudinit import CloudInitProvisionHandler, cloud_init_is_enabled
 
-def get_provision_handler(distro_name=DISTRO_NAME, 
+def get_provision_handler(distro_name=DISTRO_NAME,
                             distro_version=DISTRO_VERSION,
                             distro_full_name=DISTRO_FULL_NAME):
 
-    if conf.get_provision_cloudinit():
+    provisioning_agent = conf.get_provisioning_agent()
+
+    if provisioning_agent == 'cloud-init' or (
+            provisioning_agent == 'auto' and
+            cloud_init_is_enabled()):
+        logger.info('Using cloud-init for provisioning')
```

Comment:
I don't think this is sufficient for testing, but I may be missing a detail about how we get here. Let's say the image is built with `auto`. When the instance is running the user goes and fiddles and sets the configuration to `cloud-init` or `waagent` and then reboots. Bang!

Since neither tool knows about the the others cached data the instance is going to be provisioned from scratch, i.e. new ssh keys, hostname, etc. From my perspective, albeit the user did something they shouldn't have, this will lead to a support call.

I think somewhere along the line we need to cache the configuration and then complain when it changes on an instance that gets rebooted. This could be handled by writing some data to `/var/run/azure` and both the agent and the Azure cloud-init data source check that location. Both complain respectively if the instance on boot is not in the expected state.

Reply:
Thanks for the review! This shouldn't be an issue, though, as the provisioning handler (both of them) check to see if provisioning is already complete. So the code path should short-circuit regardless in this case, no matter which provisioning agent originally did the provisioning work.

Reply:
Regarding the question about the safety of using systemctl in a unit, from my research this should be fine. systemctl talks to systemd through a private socket, `/run/systemd/private`. See `man init` for more information on the socket itself. And [/run is loaded early on in the boot process](https://www.freedesktop.org/wiki/Software/systemd/InitrdInterface/), so the accessibility of this socket should be guaranteed in our systemd unit.